feat/re-allow multiple workers

bmiguel-teixeira opened the pull request:
#### Description

This MR does the following:

* Re adds the ability to allow multiple workers in this exporter due to: 

1. Out of Order is no longer an issue now that it is fully supported in Prometheus. Nonetheless, I am setting the default worker as 1 to avoid OoO in Vanilla Prometheus Settings.

2. With a single worker, and for a collector with a large load, this becomes "blocking". Example: Imagine a scenario in which a collector is collecting lots of targets, and with a slow prometheus/unstable network, a single worker can easily bottleneck the off-shipping if retries are enabled.

#### Link to tracking issue
N/A

#### Testing  ####

#### Documentation
docs auto-updated. Readme.md is now correct in its explanation of the `num_consumers since its no longer hard-coded at 1.  Additional docs added.

dashpole commented:
I'll be OOO for the next few weeks, so I won't be able to review until then.

ArthurSens reviewed (commented):
Thanks for the PR @bmiguel-teixeira! The contributions sound solid, but I'm a bit concerned that we're doing two separate things in one single PR here. Generally that's not a good practice because in case we have to revert one thing, we end up reverting more than what's needed, not to mention that it makes the PR biggers and a bit harder to review

Could you choose only one new functionality for this PR and open another one for the other?

--- 

Regarding the changes, could you add tests for the telemetry added? Send PRW requests to a mock server and assert that the metric you've added increments as you expected.

For allowing multiple workers, it would be nice if we add extra documentation making it clear that Out-of-order needs to be enabled in Prometheus for it to work :)

bmiguel-teixeira commented:
hi @ArthurSens 

Removed the additional telemetry to be added in a secondary PR.  Also added a bit of docs to explain the toggle and use case. Please take a look

Cheers

ArthurSens commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -149,3 +149,20 @@ sum by (namespace) (app_ads_ad_requests_total)
+
+## Multiple Consumers and OutOfOrder
+
+**DISCLAIMER**: This snippet applies only to Prometheus, other remote write destinations using Prometheus Protocol (ex: Thanos/Grafana Mimir/VictoriaMetrics) may have different settings.
+
+By default, Prometheus tipically expects samples to be ingested sequentially, in temporal order.
+
+When multiple consumers are enabled, the temporal ordering the samples written to the target destination is not deterministic and temporal ordering can no longer be garanteed. Example: one worker may be pushing a sample for `t+30s` and a second worker may push an additional sample but for `t+15s`.
+
+Vanilla Prometheus configurations will reject these unordered samples. You will find similar error to this `Error on ingesting out-of-order samples`.
+
+For enabling multiple consumers, it is necessary to enable out-of-order support in Prometheus.
+This can be done by using the `tsdb.out_of_order_time_window: 10m` settings. Please choose an appropriate time window to support pushing the worst case scenarios of a "queue" build up on the sender side.
```
```suggestion
**DISCLAIMER**: This snippet applies only to Prometheus, other remote write destinations using Prometheus Protocol (ex: Thanos/Grafana Mimir/VictoriaMetrics) may have different settings.

By default, Prometheus expects samples to be ingested sequentially, in temporal order.

When multiple consumers are enabled, the temporal ordering of the samples written to the target destination is not deterministic, and temporal ordering can no longer be guaranteed. For example, one worker may push a sample for `t+30s`, and a second worker may push an additional sample but for `t+15s`.

Vanilla Prometheus configurations will reject these unordered samples and you'll receive "out of order" errors.

Out-of-order support in Prometheus must be enabled for multiple consumers.
This can be done by using the `tsdb.out_of_order_time_window: 10m` settings. Please choose an appropriate time window to support pushing the worst-case scenarios of a "queue" build-up on the sender side.
```

ArthurSens reviewed (commented):
Well, the code is simple so it does LGTM, but I'm struggling to test this.

Do you have any examples of apps/configs that will generate out of order datapoints? All attempts I've tried provide things in order so I can't be sure if this is working as expected 😅

ArthurSens commented:
(there's a linting failure)

bmiguel-teixeira commented:
Hi @ArthurSens 

Just submitted your recomendation to fix the spelling issues.

In regards to testing and simulating locally the out of order issues.
Here is my setup.

Prometheus Config
```
global:
  scrape_interval: 1s 
  evaluation_interval: 1s

#storage:
#  tsdb:
#    out_of_order_time_window: 10m
```

Otel Config
```
receivers:
  prometheus:
    config:
      scrape_configs:
      - job_name: 'node-1'
        scrape_interval: 1s
        static_configs:
          - targets: ['127.0.0.1:8081']

exporters:
  prometheusremotewrite:
    endpoint: http://localhost:9090/api/v1/write
    remote_write_queue:
      enabled: true
      num_consumers: XXX
      queue_size: 1000
    retry_on_failure:
      enabled: true
      initial_interval: 1s
      max_interval: 5s
      max_elapsed_time: 30s

service:
  telemetry:
    logs:
      level: "DEBUG"

  pipelines:
    metrics:
      receivers: [prometheus]
      processors: []
      exporters: [prometheusremotewrite]

```

Scenario 1: (Vanilla)
* Prometheus with NO out of order window.
* PrometheusRemoteWrite with 1 consumer (or queue disabled)
1. Setup environment (let it run few couple minutes)
2. Stop Prometheus
3. Wait couple seconds for timeout errors
4. Restart Prometheus (ensure you still have old data)

Outcome: Metrics are reingested, in order with single worker, ALL GOOD.


Scenario 2 (PrometheusRemoteWrite 5 Consumers + Vanilla Prometheus)
* Prometheus with NO out of order window.
* PrometheusRemoteWrite with 5 consumers
1. Setup environment (let it run few couple minutes)
2. Stop Prometheus
3. Wait couple seconds for timeout errors (and for items to build up in the queue)
4. Restart Prometheus (ensure you still have old data)

Outcome: OutOfOrder Errors in Prometheus after boot up, since samples will be re-tried with no specific order.

Scenario 3 (PrometheusRemoteWrite 5 Consumers + Prometheus with OutOfOrder Window)
* Prometheus with WITH out of order window set at 10m
* PrometheusRemoteWrite with 5 consumers
1. Setup environment (let it run few couple minutes)
2. Stop Prometheus
3. Wait couple seconds for timeout errors (and for items to build up in the queue)
4. Restart Prometheus (ensure you still have old data)

Outcome: No OutOfOrder issues because even though we have multiple workers, Prometheus can accept mixed/old samples and update tsdb until 10 minutes "ago".


Let me know if you need any more info.

Cheers.

ArthurSens reviewed (commented):
Perfect; thank you for the instructions! I just tested it, and it works perfectly. There's just one CI failure that we need to fix before approving here.

Could you run `make generate` and commit the changes?

bmiguel-teixeira commented:
@ArthurSens All done!

dashpole commented on exporter/prometheusremotewriteexporter/factory.go:
```diff
@@ -96,7 +90,7 @@ func createDefaultConfig() component.Config {
 		RemoteWriteQueue: RemoteWriteQueue{
 			Enabled:      true,
 			QueueSize:    10000,
-			NumConsumers: 5,
```
I believe NumConsumers is used for concurrent requests here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/83f1fc67bd14584583646512748726047997d02f/exporter/prometheusremotewriteexporter/exporter.go#L259

Seems like this would decrease parallelism in the default case. Is there a way for us to do this in a way that is less likely to be breaking?  Maybe set the export concurrency to `max(5, NumConsumers)`, so we don't reduce parallelism in that part of the exporter?

bmiguel-teixeira replied:
Unless i was reading it wrong, you maintain the same parallelism anyway.
Since while the default was at 5, it ALWAYS overridden in L64 to 1.
Now I am just setting the default to 1. And the concurrency will remain the same min(1, len(requests)).
You never had concurrency it in the first place - what you had was a simple chunk logic, to split a large request into smaller requests and send them sequentially.

Now you have proper parallelism since its no longer hard-coded to 1.

Previous:
Consumers 5 (hard-coded at 1), len(requests) at 2 = Parallelism 1


Now:
Consumers 1 (default), len(requests) at 2 = Parallelism 1
Consumers 5 (default), len(requests) at 2 = Parallelism 2
Consumers 5 (default), len(requests) at 10 = Parallelism 5

dashpole replied:
The value is passed to newPRWExporter on line 45 above: `prwe, err := newPRWExporter(prwCfg, set)`.  This happens before we override NumConsumers on line 64.

My concerns are:

* Users are currently setting NumConsumers with the expectation that it will increase the number of chunks used.  E.g. users might set it to 20 to split requests more.  When we make this change, it will result in them sending with actual parallelism.
* Users can currently configure how requests are split, while still preventing out of order requests.  With this new design, they have to configure both together.

WDYT about having a separate configuration option for the request parallelism vs the number of consumers?

bmiguel-teixeira replied:
Got it. Let me double check.

I am also ok with creating an additional configuration field (and setting at 5 to keep it default as is), but I am concerned with eventual "semantic confusion" of the fields (keeping num_consumers at 5, while creating another field to control parallelism), unless we make a breaking change in the configuration.

To keep compatibility, but it kinda goes against the `exporterhelper\queue` standard that "kinda" use here.
```
another_field_worker_count: X # this controllers "workers" count.
remote_write_queue:
  num_consumers: 5 # this controls individual request chunk split
```

Ideally, maybe something like this, but a breaking change in config.
```
max_batch_size_requests: 5 # we define this field to control chunk split, since its split by the `max_batch_size_bytes`
remote_write_queue:
  num_consumers: X  # also starts getting aligned with OTLP exporterhelper
```

wdyt?

dashpole replied:
The second was what I had in mind.  We would need to feature-gate the change, but I think it is the right eventual direction.

bmiguel-teixeira replied:
awesome.
im very okay with a feature gate to move towards standardization.

Ill push the change here in the next days. 

Cheers.

bmiguel-teixeira replied:
hi @dashpole @ArthurSens 

reset changes to be behind featuregate, and updated docs. Please take a look

dashpole commented on exporter/prometheusremotewriteexporter/config.go:
```diff
@@ -87,6 +90,10 @@ var _ component.Config = (*Config)(nil)
 
 // Validate checks if the exporter configuration is valid
 func (cfg *Config) Validate() error {
+	if cfg.MaxBatchRequestParallelism <= 0 {
```
Seems like 0 should also not be valid?

bmiguel-teixeira replied:
yeah. nice catch

dashpole commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -120,13 +120,18 @@ func newPRWExporter(cfg *Config, set exporter.Settings) (*prwExporter, error) {
+	if enableMultipleWorkersFeatureGate.IsEnabled() {
+		concurrency = cfg.MaxBatchRequestParallelism
+	}
+
```
It would be nice to always always use `cfg.MaxBatchRequestParallelism`, if it is set by the user.  That way, existing users can migrate from `RemoteWriteQueue.NumConsumers` to `cfg.MaxBatchRequestParallelism` right away.  If the feature gate is disabled, and a user has set `NumConsumers` to a non-default value, it would be nice to emit a warning instructing them to migrate.  WDYT?

bmiguel-teixeira replied:
done. please check if you agree

ArthurSens commented:
@edma2 found a bug when batching time series with multiple workers. Here is a PR trying to fix it: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36524

Should we solve the bug before re-allowing multiple workers?

Aneurysm9 commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -101,12 +103,22 @@ Several helper files are leveraged to provide additional capabilities automatica
+This exporter has feature gate: `+exporter.prometheusremotewritexporter.EnableMultipleWorkers`.
+
+When this feature gate is enabled, `num_consumers` will be used as the worker counter for handling batches from the queue, and `max_batch_request_parallelism` will be used for parallelism on single batch bigger than `max_batch_size_bytes`.
+Enabling this feature gate, with `num_consumers` higher than 1 requires the target destination to supports ingestion of OutOfOrder samples. See [Multiple Consumers and OutOfOrder](#multiple-consumers-and-outoforder) for more info
```
I think this value needs to default to `1` if this feature gate is enabled.  When the feature gate is ultimately removed I don't think it appropriate to have a default value that is known to not work with any number of potential receivers and would not work with the default configuration of those receivers that are known to be able to support it in some configurations.

bmiguel-teixeira replied:
I aggree that will help with transition, and that the exporter should have "sane" defaults at all times. 
done. please check

dashpole commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -120,13 +120,18 @@ func newPRWExporter(cfg *Config, set exporter.Settings) (*prwExporter, error) {
 
 	userAgentHeader := fmt.Sprintf("%s/%s", strings.ReplaceAll(strings.ToLower(set.BuildInfo.Description), " ", "-"), set.BuildInfo.Version)
 
+	concurrency := cfg.RemoteWriteQueue.NumConsumers
+	if enableMultipleWorkersFeatureGate.IsEnabled() || cfg.MaxBatchRequestParallelism != nil {
+		concurrency = *cfg.MaxBatchRequestParallelism
+	}
```
It will only be a change in behavior once the feature gate is enabled.  It shouldn't be a breaking change in this PR (unless i'm mistaken). The reason we have a feature gate at all is to make the migration from the breaking change easier.

bmiguel-teixeira replied:
i missed the gate difference. ill update this over the weekend

bmiguel-teixeira replied:
Updated. Please check

ArthurSens commented:
Hey @bmiguel-teixeira, once https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36601 is merged, I think we should be safe to proceed with multiple workers again :)

Could you rebase your PR on top of main once that happens?

ArthurSens commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -149,3 +161,20 @@ sum by (namespace) (app_ads_ad_requests_total)
+
+See for more info:
+- https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb
+- https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver
```
Could we remove it?

bmiguel-teixeira replied:
upstream doc may have hanged. will remove this entry.

bmiguel-teixeira replied:
done

ArthurSens reviewed (commented):
While reading the changes in README, I'm confused about the difference between `num_consumers` and `max_batch_request_parallelism`. 

Both mention that they configure the number of workers... so they fight each other to decide the number of workers? Do they complement each other?

--- 

By the way, thanks for the patience here, a lot was done to make this possible :)

ArthurSens reviewed (approved):
The README is still a bit confusing to me, but I'm fine merging if the other codeowners are ok with it

bmiguel-teixeira commented:
Hi @ArthurSens 

No worries on the delays.
The plan, as discussed with @dashpole, is to start the process of transitioning out of the hard-coded values and into the ability to configure both workers and single request parallelism (for large payloads of single-requests) individually.

If `max_batch_request_parallelism` is defined, this is the value used for single-requests parallelism regardless of anything else. Users can start moving to this value starting from this PR to keep same behaviour.

Then we have feature gate for controlling the parallelism. 
If `MultipleWorkersFeatureGate` is enabled, then `numConsumers` enables multiple workers/consumers from the metrics queue, and single request parallelism can be set with `max_batch_request_parallelism`.

If the gate is absent, we keep the default behaviour with single worker thread and with `numConsumer` controlling single request parallelism.

Hopefully this helps

Aneurysm9 commented on exporter/prometheusremotewriteexporter/README.md:
```diff
@@ -66,6 +66,8 @@ The following settings can be optionally configured:
   samples to be sent to the remote write endpoint. If the batch size is larger
   than this value, it will be split into multiple batches.
+- `max_batch_request_parallelism` (default = `5`): Maximum parallelism allowed for a single request bigger than `max_batch_size_bytes`.
+  This configuration is used only when feature gate `exporter.prometheusremotewritexporter.EnableMultipleWorkers` is enabled.
```
I think the implementation differs from this documentation.  It seems like this config option will be used if set regardless of the state of the feature gate.

dashpole replied:
@Aneurysm9 is correct.  Please remove this sentence

bmiguel-teixeira replied:
Correct. I didnt review README after all the changes. Please take another look. If it LGTM do let me know to squash and rebase

MovieStoreGuy commented on exporter/prometheusremotewriteexporter/config_test.go:
```diff
@@ -136,3 +141,8 @@ func TestDisabledTargetInfo(t *testing.T) {
 	assert.False(t, cfg.(*Config).TargetInfo.Enabled)
 }
+
+func toPtr[T any](val T) *T {
+	ret := val
+	return &ret
+}
```
Go is pass by copy, so it is fine to take the address of the val

bmiguel-teixeira replied:
ur right. updated.

andrzej-stencel commented:
Looks like the linter is not happy https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12754053453/job/35549100453?pr=36134:

> Error: dorisexporter/exporter_common.go:127:6: var-naming: func toJsonLines should be toJSONLines (revive)

songy23 commented:
Please rebase

songy23 commented:
> Generated code is out of date, please apply this diff and commit the changes in this PR.

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12830166716/job/35777750399?pr=36134

please fix

bmiguel-teixeira committed:
feat/re-allow multiple workers

songy23 commented:
NVM, the errors in code owners are from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37279 this PR should be good

bmiguel-teixeira commented:
@songy23 let me know if anything else is needed. Cheers
